server/pbench/bin/pbench-dispatch

```diff
@@ -65,102 +65,108 @@ typeset -i ndups=0
 
 # the link source and destination for this script
 linksrc=TODO
-linkdestlist=$(getconf.py -l dispatch-states pbench-server)
-if [[ -z "$linkdestlist" ]]; then
-    echo "config file error: either no dispach-states defined or a typo" >&4
-    subj="$PROG.$TS($PBENCH_ENV) - Config file error"
-    echo "config file error: either no dispach-states defined or a typo" | mailx -s "$subj" $mail_recipients
-    exit 1
-else
-    echo $TS
-
-    # get the list of files we'll be operating on - sort them by size
-    list=$TMP/pbench-unpack-tarballs.$$
-    find -L $ARCHIVE/*/$linksrc -name '*.tar.xz' 2>/dev/null | grep -v DUPLICATE > $list
-
-    trap 'rm -f $list' EXIT INT QUIT
-
-    while read result ;do
-        ntotal=$ntotal+1
-
-        link=$(readlink -e $result)
-        if [ ! -f "$link" ] ;then
-            echo "$TS: $link does not exist" >&4
-            nerrs=$nerrs+1
-            continue
-        fi
-
-        resultname=$(basename $result)
-        resultname=${resultname%.tar.xz}
-        hostname=$(basename $(dirname $link))
-
-        # echo $link
-        # echo $resultname
-        # echo $hostname
-        # continue
-
-        # make sure that all the relevant state directories exist
-        mk_dirs $hostname
-        # ... and a couple of other necessities.
-        if [[ $? -ne 0 ]] ;then
-            echo "$TS: Creation of $hostname processing directories failed: code $status" >&4
-            nerrs=$nerrs+1
-            continue
-        fi
-        
-        mkdir -p $TMP/$PROG/$hostname
-        status=$?
-        if [[ $status -ne 0 ]] ;then
-            echo "$TS: mkdir -p $TMP/$PROG/$hostname failed: code $status" >&4
-            nerrs=$nerrs+1
-            continue
-        fi
-        
-        # XXXX - for now, if it's a duplicate name, just punt and avoid producing the error - the full
-        # solution will involve renaming the unpacked directory appropriately.
-        if [ ${resultname%%.*} == "DUPLICATE__NAME" ] ;then
-            ndups=$ndups+1
-            continue
-        fi
 
-        # check the md5 sum
-        md5=$(cat $link.md5 2> /dev/null | cut -d' ' -f1)
-        md5tar=$(md5sum $link 2> /dev/null | cut -d' ' -f1)
-        if [ "$md5" != "$md5tar" ] ;then
-            echo "$TS: MD5 check failed on $link" >&4
-            # move the link so that the script does not report the error continuously
-            mv $result $(echo $result | sed 's/TODO/BAD-MD5/')
-            nerrs=$nerrs+1
-            continue
+monitor_dir=$ARCHIVE/*/$linksrc
+inotifywait -m $monitor_dir -e create,moved_to | while read path action file; do
```

Comment:
We'll need to pass the list of monitored directories from a file using `--fromfile $file` or `--fromfile -` and passing them through stdin. We might have enough hosts already to blow the command line length limit.

We'll also need to add a watch for a newly-created `$ARCHIVE/foo` directory, when previously unknown host `foo` sends its first result over. The action for that should be to add `$ARCHIVE/foo/$linksrc` to the watch list.